From d86be15cc4230a7fce43b32fdfc4b741fd6caa10 Mon Sep 17 00:00:00 2001
From: Richard Lowe <richlowe@richlowe.net>
Date: Sat, 27 Oct 2012 02:44:09 +0100
Subject: [PATCH 05/12] Implement -fstrict-calling-conventions

Stock GCC is overly willing to violate the ABI when calling local functions,
such that it passes arguments in registers on i386.  This hampers debugging
with anything other than a fully-aware DWARF debugger, and is generally not
something we desire.

Implement a flag which disables this behaviour, enabled by default.  The flag is
global, though only effective on i386, to more easily allow its globalization
later which, given the odds, is likely to be necessary.
diff -wpruN '--exclude=*.orig' gcc-8.1.0~/gcc/common.opt gcc-8.1.0/gcc/common.opt
--- gcc-8.1.0~/gcc/common.opt	2018-03-28 14:51:09.000000000 +0000
+++ gcc-8.1.0/gcc/common.opt	2018-05-24 11:10:01.513110905 +0000
@@ -2414,6 +2414,10 @@ fstrict-aliasing
 Common Report Var(flag_strict_aliasing) Optimization
 Assume strict aliasing rules apply.
 
+fstrict-calling-conventions
+Common Report Var(flag_strict_calling_conventions) Init(1)
+Use strict ABI calling conventions even for static functions
+
 fstrict-overflow
 Common Report
 Treat signed overflow as undefined.  Negated as -fwrapv -fwrapv-pointer.
diff -wpruN '--exclude=*.orig' gcc-8.1.0~/gcc/config/i386/i386.c gcc-8.1.0/gcc/config/i386/i386.c
--- gcc-8.1.0~/gcc/config/i386/i386.c	2018-04-25 17:31:20.000000000 +0000
+++ gcc-8.1.0/gcc/config/i386/i386.c	2018-05-24 11:10:01.523151611 +0000
@@ -6730,6 +6730,7 @@ ix86_function_regparm (const_tree type,
 	 and callee not, or vice versa.  Instead look at whether the callee
 	 is optimized or not.  */
       if (target && opt_for_fn (target->decl, optimize)
+	  && !flag_strict_calling_conventions
 	  && !(profile_flag && !flag_fentry))
 	{
 	  cgraph_local_info *i = &target->local;
@@ -6827,6 +6828,7 @@ ix86_function_sseregparm (const_tree typ
       /* TARGET_SSE_MATH */
       && (target_opts_for_fn (target->decl)->x_ix86_fpmath & FPMATH_SSE)
       && opt_for_fn (target->decl, optimize)
+	  && !flag_strict_calling_conventions
       && !(profile_flag && !flag_fentry))
     {
       cgraph_local_info *i = &target->local;
diff -wpruN '--exclude=*.orig' gcc-8.1.0~/gcc/doc/invoke.texi gcc-8.1.0/gcc/doc/invoke.texi
--- gcc-8.1.0~/gcc/doc/invoke.texi	2018-04-26 13:52:59.000000000 +0000
+++ gcc-8.1.0/gcc/doc/invoke.texi	2018-05-24 11:10:01.528691725 +0000
@@ -9185,6 +9185,12 @@ int f() @{
 The @option{-fstrict-aliasing} option is enabled at levels
 @option{-O2}, @option{-O3}, @option{-Os}.
 
+@item -fstrict-calling-conventions
+@opindex fstrict-calling-conventions
+Use strict ABI calling conventions even with local functions.
+This disable certain optimizations that may cause GCC to call local
+functions in a manner other than that described by the ABI.
+
 @item -falign-functions
 @itemx -falign-functions=@var{n}
 @opindex falign-functions
diff -wpruN '--exclude=*.orig' gcc-8.1.0~/gcc/testsuite/gcc.target/i386/local.c gcc-8.1.0/gcc/testsuite/gcc.target/i386/local.c
--- gcc-8.1.0~/gcc/testsuite/gcc.target/i386/local.c	2015-12-29 10:32:21.000000000 +0000
+++ gcc-8.1.0/gcc/testsuite/gcc.target/i386/local.c	2018-05-24 11:10:01.528974387 +0000
@@ -1,5 +1,6 @@
 /* { dg-do compile } */
-/* { dg-options "-O2 -funit-at-a-time" } */
+/* { dg-options "-O2 -funit-at-a-time -fno-strict-calling-conventions" { target ia32 } } */
+/* { dg-options "-O2 -funit-at-a-time" { target lp64 } } */
 /* { dg-final { scan-assembler "magic\[^\\n\]*eax" { target ia32 } } } */
 /* { dg-final { scan-assembler "magic\[^\\n\]*(edi|ecx)" { target { ! ia32 } } } } */
 
diff -wpruN '--exclude=*.orig' gcc-8.1.0~/gcc/testsuite/gcc.target/i386/strict-cc.c gcc-8.1.0/gcc/testsuite/gcc.target/i386/strict-cc.c
--- gcc-8.1.0~/gcc/testsuite/gcc.target/i386/strict-cc.c	1970-01-01 00:00:00.000000000 +0000
+++ gcc-8.1.0/gcc/testsuite/gcc.target/i386/strict-cc.c	2018-05-24 11:10:01.529200959 +0000
@@ -0,0 +1,24 @@
+/* { dg-do compile { target { ilp32 } } } */
+/* { dg-options "-O2 -funit-at-a-time -fstrict-calling-conventions"  } */
+/* { dg-final { scan-assembler "pushl.*\\\$1" } } */
+/* { dg-final { scan-assembler "pushl.*\\\$2" } } */
+/* { dg-final { scan-assembler "pushl.*\\\$3" } } */
+/* { dg-final { scan-assembler "pushl.*\\\$4" } } */
+/* { dg-final { scan-assembler "pushl.*\\\$5" } } */
+
+#include <stdio.h>
+
+/* Verify that local calling convention is not used if strict conventions.  */
+static int t(int, int, int, int, int) __attribute__ ((noinline));
+
+int
+m()
+{
+    t(1, 2, 3, 4, 5);
+}
+
+static int
+t(int a, int b, int c, int d, int e)
+{
+    printf("%d\n", a, b, c, d, e);
+}
